ledger/participant-integration-api/src/main/scala/platform/store/ActiveLedgerStateManager.scala

```diff
@@ -69,146 +100,174 @@ private[platform] class ActiveLedgerStateManager[ALS <: ActiveLedgerState[ALS]](
-                val nodeParties = nc.signatories
-                  .union(nc.stakeholders)
-                  .union(nc.key.map(_.maintainers).getOrElse(Set.empty))
-                val activeContract = ActiveContract(
-                  id = nc.coid,
-                  let = let,
-                  transactionId = transactionId,
-                  nodeId = nodeId,
-                  workflowId = workflowId,
-                  contract = nc.versionedCoinst,
-                  witnesses = disclosure(nodeId),
-                  // A contract starts its life without being divulged at all.
-                  divulgences = Map.empty,
-                  key = nc.versionedKey.map(
-                    _.assertNoCid(coid => s"Contract ID $coid found in contract key")
-                  ),
-                  signatories = nc.signatories,
-                  observers = nc.stakeholders.diff(nc.signatories),
-                  agreementText = nc.coinst.agreementText,
-                )
-                activeContract.key match {
-                  case None =>
-                    ats.copy(
-                      acc = Some(acc.addContract(activeContract, None)),
-                      parties = parties.union(nodeParties),
+              if (actAs.nonEmpty) {
+                // If the submitter is known, look up the contract
+                // Submitters being know means the transaction was submitted on this participant.
+                val currentResult = als.lookupContractByKey(gk)
+                if (currentResult == nlkup.result) {
+                  state.copy(
+                    parties = state.parties.union(nodeParties)
+                  )
+                } else {
+                  state.copy(
+                    errs = state.errs + Inconsistent(
+                      s"Contract key lookup with different results: expected [${nlkup.result}], actual [${currentResult}]"
                     )
-                  case Some(key) =>
-                    val gk = GlobalKey(activeContract.contract.template, key.key.value)
-                    if (acc.lookupContractByKey(gk).isDefined) {
-                      AddTransactionState(
-                        None,
-                        errs + Inconsistent("DuplicateKey: contract key is not unique"),
-                        parties.union(nodeParties),
-                        archivedIds,
-                      )
-                    } else {
-                      ats.copy(
-                        acc = Some(acc.addContract(activeContract, Some(gk))),
-                        parties = parties.union(nodeParties),
-                      )
-                    }
+                  )
                 }
-              case ne: N.NodeExercises[_, ContractId] =>
+              } else {
+                // Otherwise, trust that the lookup was valid.
+                // The submitter being unknown means the transaction was submitted on a different participant,
```

Comment:
Note unrelated to this PR: this class is used in two places: 
- In-memory sandbox classic. In this case, there is only a single participant and we have full trust in the daml interpretation during command execution.
- In a very old migration of the index database. In this case, we *could* encounter transactions submitted through different participants. This use prevents us from simplifying or removing this class, and we should at some point try to get rid of this migration.

Reply:
I’ve added a comment where this is used at the top of the class 51f73798e726066f8eb07c3934a0f3ea0c9b5af6